content/en/docs/tasks/kubelet-credential-provider/kubelet-credential-provider.md

```diff
@@ -0,0 +1,114 @@
+---
+title: Kubelet Image Credential Provider
+reviewers:
+- liggitt
+- cheftako
+description: Configure the kubelet's image credential provider plugin
+content_type: task
+---
+
+<!-- overview -->
+
+Starting from Kubernetes v1.20, the kubelet can dynamically retrieve credentials for a container image registry
+using exec plugins. The kubelet and the exec plugin communicate through stdio (stdin, stdout, and stderr) using
+Kubernetes versioned APIs. These plugins allow the kubelet to request credentials for a container registry dynamically
+as oppposed to storing static credentials on disk. For example, the plugin may talk to a local metadata server to retrieve
+short-lived credentials for an image that is being pulled by the kubelet.
+
+You may be interested in using this capabilitiy if any of the below are true:
+1. API calls to a cloud provider service are required to retreive authentication information for a registry.
+2. Credentials have short expiration times and requesting new credentials frequently is required.
+3. Storing registry credentials on disk or in image pull Secrets is not acceptable.
+
+This guide demonstrates how to configure the kubelet's image credential provider plugin mechanism.
+
+{{< note >}}
+This capabilitiy will eventually replace built-in logic in the kubelet to serve registry credentials for Azure, AWS and GCP.
+{{< /note >}}
+
+## {{% heading "prerequisites" %}}
+
+* kubelet on v1.20 or later. If on v1.20, ensure the alpha feature gate `KubeletCredentialProviders` is on.
+* A working implementation of a credential provider exec plugin. You can build your own plugin or use one provided by cloud providers.
+
+<!-- steps -->
+
+## Installing Plugins on Nodes
+
+A credential provider plugin is an exectuable binary that will be run by the kubelet. Ensure that the plugin binary exists on
+every node in your cluster and stored in a known directory. The directory will be required later when configuring kubelet flags.
+
+## Configuring Kubelet
+
+### Flags
+
+In order to use this feature, the kubelet expects two flags to be set:
+* `--image-credential-provider-config` - the path to the credential provider plugin config file.
+* `--image-credential-provider-bin-dir` - the path to the directory where credential provider plugin binaries are located.
+
+If you are on v1.20, ensure the feature gate `KubeletCredentialProviders` is enabled via the `--feature-gates` flag.
+
+### Credential Provider Configuration File
+
+The configuration file passed into `--image-credential-provider-config` is read by the kubelet to determine which exec plugins
+should be invoked for which container images. Here's an example configuration file you may end up using if you are using the ECR-based plugin:
+
+```yaml
+kind: CredentialProviderConfig
```

Comment:
It is a separate API group, distinct from kubelet config https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/kubelet/pkg/apis/credentialprovider

Reply:
@andrewsykim The API group you are talking about only contains `CredentialProviderRequest`, `CredentialProviderRequest` and `AuthConfig`. It does not contain `CredentialProviderConfig`. I was hoping that `CredentialProviderConfig` in the `k8s.io/kubelet/config/v1beta1/types.go`.

Reply:
Oh my bad, CredentialProviderConfig is in the v1alpha1 kubelet config API group for v1.20.

Reply:
The weird thing is that `CredentialProviderConfig` only exists in https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/types.go#L468, not in the staging repo: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/config/v1beta1/types.go

Is this a bug to fix?

I noticed this because I was trying to generate kubelet configuration documentation, along with other "unpublished" API types. See: https://github.com/kubernetes/website/pull/25094 and the preview is here: https://deploy-preview-25094--kubernetes-io-master-staging.netlify.app/docs/reference/command-line-tools-reference/kubelet-config.v1beta1/
`CredentialProviderConfig` is not shown on that page.

Reply:
`CredentialProviderConfig` is in staging, but it is only in the v1alpha1 API group https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kubelet/config/v1alpha1/types.go. It will be promote to the v1beta1 group in a future release.

Reply:
Ah I see. Sorry for all the stupid questions. I was thinking that "if everything is up to date, it should appear in the latest version", however, that is not the case. Thanks for the clarification.